src/lib/hash-parser-hoc.jsx

```diff
@@ -13,35 +21,70 @@ const HashParserHOC = function (WrappedComponent) {
                 'handleHashChange'
             ]);
             this.state = {
-                projectId: null
+                hideIntro: false
             };
         }
         componentDidMount () {
             window.addEventListener('hashchange', this.handleHashChange);
             this.handleHashChange();
         }
+        componentDidUpdate (prevProps) {
+            // if we are newly fetching a non-hash project...
+            if (this.props.isFetchingProjectWithNoURLId && !prevProps.isFetchingProjectWithNoURLId) {
+                // ...clear the hash from the url
+                history.pushState('new-project', 'new-project',
+                    window.location.pathname + window.location.search);
+            }
+        }
         componentWillUnmount () {
             window.removeEventListener('hashchange', this.handleHashChange);
         }
         handleHashChange () {
             const hashMatch = window.location.hash.match(/#(\d+)/);
-            const projectId = hashMatch === null ? 0 : hashMatch[1];
-            if (projectId !== this.state.projectId) {
-                this.setState({projectId: projectId});
+            const hashProjectId = hashMatch === null ? defaultProjectId : hashMatch[1];
+            if (hashProjectId !== this.props.reduxProjectId) {
+                this.props.setHashProjectId(hashProjectId);
+            }
+            if (hashProjectId !== defaultProjectId) {
+                this.setState({hideIntro: true});
             }
         }
         render () {
+            const {
+                /* eslint-disable no-unused-vars */
+                isFetchingProjectWithNoURLId: isFetchingProjectWithNoURLIdProp,
+                reduxProjectId,
+                setHashProjectId: setHashProjectIdProp,
+                /* eslint-enable no-unused-vars */
+                ...componentProps
+            } = this.props;
             return (
                 <WrappedComponent
-                    hideIntro={this.state.projectId && this.state.projectId !== 0}
-                    projectId={this.state.projectId}
-                    {...this.props}
+                    hideIntro={this.state.hideIntro}
+                    {...componentProps}
                 />
             );
         }
     }
-
-    return HashParserComponent;
+    HashParserComponent.propTypes = {
+        isFetchingProjectWithNoURLId: PropTypes.bool,
+        reduxProjectId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+        setHashProjectId: PropTypes.func
+    };
+    const mapStateToProps = state => {
+        const projectState = state.scratchGui.projectId.projectState;
+        return {
+            isFetchingProjectWithNoURLId: isFetchingProjectWithNoURLId(projectState),
```

Comment:
This is confusing because I thought `isFetchingProjectWithNoURLId` was a boolean prop, but here it's a function that returns a boolean based on `projectState` — I think the function name probably could indicate that it calculates the value rather than having the same name as the boolean.

It's also not clear what `projectState` is — the state of the project could be the state of the currently running project, etc. But it seems like it's the loading state, so I think having `loading` in the name would help (if that's true!)

Moreover if all of this comes from the `projectId` reducer already, then couldn't the `isFetchingProjectWithNoURLId` prop (or something with a different name that gives the same information) be stored in the state without having to call a function to return it?

Reply:
Ok, the more I think about calling the state property `loadingState`, the more I like it...

I renamed `isFetchingProjectWithNoURLId` to `getIsFetchingWithoutId`, and renamed other selectors accordingly. I think it's much clearer.